libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,208 @@
+#include <stdbool.h>
+
+#include "adc.h"
+#include "log.h"
+#include "stm32f0xx.h"
+
+#define TS_CAL1 0x1ffff7b8
+#define TS_CAL2 0x1ffff7c2
+#define VREFINT_CAL 0x1ffff7ba
+
+typedef struct ADCInterrupt {
+  ADCCallback callback;
+  void *context;
+  uint16_t reading;
+} ADCInterrupt;
+
+typedef struct ADCStatus {
+  ADCChannel current_channel;
+  uint32_t select;
+  bool continuous;
+} ADCStatus;
+
+static ADCInterrupt s_adc_interrupts[NUM_ADC_CHANNEL];
+static ADCStatus s_adc_status;
+
+static int16_t prv_get_temp(int32_t reading) {
+  uint16_t ts_cal1 = *(uint16_t*)TS_CAL1;
+  uint16_t ts_cal2 = *(uint16_t*)TS_CAL2;
+
+  reading = (110 - 30) * (reading - ts_cal1) / (ts_cal2 - ts_cal1) + 30;
+
+  return reading + 273;
+}
+
+static uint16_t prv_get_vdda(uint32_t reading) {
+  uint16_t vrefint_cal = *(uint16_t*)VREFINT_CAL;
+  reading = (3300 * vrefint_cal) / reading;
+  return reading;
+}
+
+void adc_init(ADCMode adc_mode) {
+  // Stop ongoing conversions and disable the ADC
+  ADC_StopOfConversion(ADC1);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADSTP)) { }
+
+  ADC1->CR |= ADC_FLAG_ADDIS;
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADEN)) { }
+
+  // Once the ADC has been reset, enable it with the given settings
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    .ADC_Resolution = ADC_Resolution_12b,
+    .ADC_ContinuousConvMode = adc_mode,
+    .ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None,
+    .ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_TRGO,
+    .ADC_DataAlign = ADC_DataAlign_Right,
+    .ADC_ScanDirection = ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+  ADC_TempSensorCmd(ENABLE);
+  ADC_VrefintCmd(ENABLE);
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  // Set the ADC to wait mode
+  ADC_WaitModeCmd(ADC1, ENABLE);
+  ADC_AutoPowerOffCmd(ADC1, ENABLE);
```

Comment:
We might only want to enable auto power off if we're running at a low frequency (single mode or slower conversions), otherwise continuously switching the ADC on and off might end up taking longer and consume more power.

![image](https://user-images.githubusercontent.com/5583338/27664408-6697a730-5c35-11e7-94b0-21cfb2bca131.png)